analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt

```diff
@@ -33,6 +33,7 @@ import org.ossreviewtoolkit.analyzer.managers.utils.PackageManagerDependencyHand
 import org.ossreviewtoolkit.analyzer.managers.utils.SpdxDocumentCache
 import org.ossreviewtoolkit.analyzer.managers.utils.SpdxResolvedDocument
 import org.ossreviewtoolkit.downloader.VersionControlSystem
+import org.ossreviewtoolkit.model.CuratedPackage
```

Comment:
I have the following thoughts about this change which is why I tend towards not agreeing

pros:
1. Step towards cleanly separating legal conclusion from package meta-data.
  - If this is just a step, what is the end goal? What other steps are planned.
  - Is it really more separation, as there has already been a dedicated property which IMO achives separation already.
  - `CuratedPackage` is not just about legal conclusions, but about curations in general. It may curate any field from     `Package`, not just legal conclusions. So, why should this class be used to separate legal conclusions?

cons:
1. Most of the code does not care about which curations have been applied, but just about curated package meta-data.
    All that code now becomes more complicated / verbose, including the test.
3. Code has more dependencies than needed and so becomes less flexible and harder to test.
4. Writing code which is agnostic of whether the input data is curated or not becomes harder / impossible.
5. Other properties from the package class may also get `curated` so `CuratedPackage` is not just about legal   conclusion
6. Law of Demeter gets violated more often
7. I don't see any problem with having all attributes for a `Package` in the `Package` class. It seems reasonable as these are closely related and usually needed altogether.

What do you think about that? 
Also, further opinions would be interresting, e.g. @oss-review-toolkit/core-devs ?

Reply:
> If this is just a step, what is the end goal? What other steps are planned?

Ultimately, I'd like what I start to call "metadata corrections" (i.e. changes required to package metadata in order for the downloader to be able to download the source code) to also be configured in a different file than "legal conclusions" (i.e. concluding licenses or copyrights). That's because in most real-world workflows I've seen so far, these very different kinds of "curations" are performed by different people, and ORT should do a better job at providing role-specific configuration.

> Is it really more separation, as there has already been a dedicated property which IMO achives separation already.

With the above in mind, that's not enough separation. As we (de-)serialize on a per-class basis, things that should ultimately go to different files also need to go to different classes.

> `CuratedPackage` is not just about legal conclusions, but about curations in general.

That's exactly the problem I'd like to solve. IMO it has been a severe data model design flaw to allow curating "everything" (even semantically totally unrelated stuff) via a single curations mechanism.

Reply:
> With the above in mind, that's not enough separation. As we (de-)serialize on a per-class basis, things that should ultimately go to different files also need to go to different classes.

Yes, that's right. However, that argumentation does not apply as the data models changed in this PR are actually not part of ORT's configuration IIUC. So, these classes do not affect the way curations are written. As these classes are not part of the configuration, I believe the design discussion can remain independent of ORT's configuration interface.

Reply:
> the data models changed in this PR are actually not part of ORT's configuration

They are not part of ORT's *configuration*, right, but `CuratedPackage` is serialized as part of ORT's results. And IMO the mentioned (planned) separation makes sense to be also reflected there. For the configuration, *way* more refactoring would be needed, and I didn't want to dive into that rabbit hole just now before find a general agreement on the way forward.

Reply:
I still have problems following the argumentation, let me try to explain in other words what I still believe I don't get yet:

> Ultimately, I'd like what I start to call "metadata corrections" (i.e. changes required to package metadata in order for the downloader to be able to download the source code) to also be configured in a different file than "legal conclusions" (i.e. concluded licenses or copyrights). That's because in most real-world workflows I've seen so far, these very different kinds of "curations" are performed by different people, and ORT should do a better job at providing role-specific configuration.

This section is about ORT's configuration interface, but this PR is not about ORT's configuration interface.

> With the above in mind, that's not enough separation. As we (de-)serialize on a per-class basis, things that should ultimately go to different files also need to go to different classes.

This again is about ORT's configuration interface, but this PR is not about ORT's configuration interface.

 > That's exactly the problem I'd like to solve. IMO it has been a severe data model design flaw to allow curating "everything" (even semantically totally unrelated stuff) via a single curations mechanism.

This PR does not change any curation mechanism. So, there is still a single one.

I'd like to understand what exactly you mean by this "severe design" flow.
Is it in ORT's configuration interface, or in the OrtResult? 

> They are not part of ORT's configuration, right, but CuratedPackage is serialized as part of ORT's results. And IMO the mentioned (planned) separation makes sense to be also reflected there. For the configuration, way more refactoring would be needed, and I didn't want to dive into that rabbit hole just now before find a general agreement on the way forward.

What I understand is that the proposal is to separate legal conclusions from other meta-data curations in ORT's configuration interface (the user facing API), so that the different "roles" involved in the process only need looking into the config which is relevant to them.

What I still don't get is why the problem isn't solved by introducing or finding a way to separate things 
at ORT's configuration API and limiting the solution to the configuration API. So, is there need for touching the rules API as well as the OrtResult? Note: I believe historically OrtResult was intended to be human readable, but nowadays it's moving more towards a more redundancy free file format, like with the introduction of the dependency graph. Also the rules API to me seems to target more technical persons, but not all roles involved in the process. In short, why does OrtResult and rules API need to be refactored? To make it readable by legal persons?

Apart from that, have you considered using ORT's configuration API (almost) as-is to solve the problem?
I'm talking about an approach like the following, which is using a single curation mechanism but allows users to separate things if they want to.

```
curations/provenances.yml
curations/concluded-licenses.yml
curations/declared-license-mappings.yml
curations/other.yml
```

Reply:
> This section is about ORT's configuration interface, but this PR is not about ORT's configuration interface.

Correct. I drifted off-topic a bit due to your questions, trying to explain the motivation for these changes by some upcoming (IMO) related changes that *will* affect the configuration.

> I'd like to understand what exactly you mean by this "severe design" flow. Is it in ORT's configuration interface, or in the OrtResult or in the Rules API?

In all of these we have design flaws, IMO. Going into details here would probably again be off-topic, but briefly:

- In the ORT data model classes we have several problems IMO, esp. around the general topic of curated data. The `Package` cass for example should only capture package metadata as declared by the package managers. It's a superset of the metadata that a package manager providers about a package. Consequently, the `concludedLicense` should not be part of that class, as package managers do not provide a concluded license (with the exception of the SPDX "fake" package manager). Also, alone from the pure naming of the class, it's clear that the `concludedLicense` property should go to the `CuratedPackage` class. This is also backed by this [comment](https://github.com/oss-review-toolkit/ort/blob/a8ef0a09150ae5b0f304661cd82e2867a649f17e/model/src/main/kotlin/Package.kt#L81) which basically admits that the `concludedLicense` property is wrong at the current place.
- In the configuration, we have the concepts of (package) curations, package configurations, license finding curations etc. The main problem with these is that they were modeled on a very technical model around ORT's (partly flawed) data model in this area. Instead, we should have modeled the configuration after real-world use-cases and user's roles. For example, in the workflows I've seen implemented in companies so far, hardly ever the same people deal with fixing package metadata so that the downloader is able to download a package vs. making legal conclusions (e.g. by setting a concluded license). Due to these very different audiences, these things should also not be configured in the same file, so that files can more easily be edited independently.
- Regarding the rules API, we already hab talked several times in the developer meeting about it's current limitations, like assuming that every reported violation is somehow license-related, and @mnonnenmacher's refactoring plans.

> What I understand is that the proposal is to separate legal conclusions from other meta-data curations in ORT's configuration interface (the user facing API), so that the different "roles" involved in the process only need looking into the config which is relevant to them.

Correct, also see what I've written above.

> So, is there need for touching the rules API as well as the OrtResult?

Touching the rules API in this PR was just a by-product of the refactoring. When starting with the refactoring to replace usages of `Package` with `CuratedPackage` where needed, because the `concludedLicense` field has moved, I realized the independent use of `pkg` and `curations` in the rules, which just did not make any sense to me (neither before nor after my changes). Looking at the current code base only, I could have also just dropped `curations` from the rules (as I actually agree with you when you say that rules hardly access curations), but that wouldn't have worked with the upcoming changes, as rules may want to access the concluded license.

However, now that I think about it again, maybe the `curations` could *still* be dropped from the rules, as the rules would have access to the concluded license via the `ResolvedLicenseInfo`. Correct @mnonnenmacher?

> In short, why does OrtResult and rules API need to be refactored? To make it readable by legal persons?

I don't think that (purely) legal persons will every be able to write (bigger) rules, as it'll still be too complicated even after a refactoring. I mean, these people hardly ever have IDEs installed to get auto-completion etc.

But yes, we need to make writing rules easier in any case.

Regarding the `OrtResult`, I don't see *so* much need for refactoring there; but still, e.g. the things that can be curated / configured separately, should still stay separate in the `OrtResult` for easy resemblance. For example, I'd liek to be able to quickly show which consecute lines of data in the `OrtResult` are affected by a metadata correct vs. a legal conclusion.

> Apart from that, have you considered using ORT's configuration API (almost) as-is to solve the problem?

I believe you're confusing the (programmatic) API (i.e. functions) a bit with data models and / or serialized files. But in the end, yes, a split on the file basis similar to what you suggest is also on my mind. But that discussion should not be held as part of this PR, IMO.